Mathlib/Probability/Kernel/Disintegration/Density.lean

```diff
@@ -0,0 +1,799 @@
+  rw [← set_lintegral_univ]
+  exact set_lintegral_density hκν a hs MeasurableSet.univ
+
+end Integral
+
+lemma tendsto_integral_density_of_monotone (hκν : fst κ ≤ ν) [IsFiniteKernel κ] [IsFiniteKernel ν]
+    (a : α) (s : ℕ → Set β) (hs : Monotone s) (hs_iUnion : ⋃ i, s i = univ)
+    (hs_meas : ∀ n, MeasurableSet (s n)) :
+    Tendsto (fun m ↦ ∫ x, density κ ν a x (s m) ∂(ν a)) atTop (𝓝 (κ a univ).toReal) := by
+  simp_rw [integral_density hκν a (hs_meas _)]
+  have h_cont := ENNReal.continuousOn_toReal.continuousAt (x := κ a univ) ?_
+  swap
+  · rw [mem_nhds_iff]
+    refine ⟨Iio (κ a univ + 1), fun x hx ↦ ne_top_of_lt (?_ : x < κ a univ + 1), isOpen_Iio, ?_⟩
+    · simpa using hx
+    · simp only [mem_Iio]
+      exact ENNReal.lt_add_right (measure_ne_top _ _) one_ne_zero
+  refine h_cont.tendsto.comp ?_
+  have h := tendsto_measure_iUnion (s := fun n ↦ univ ×ˢ s n) (μ := κ a) ?_
+  swap; · intro n m hnm x; simp only [mem_prod, mem_univ, true_and]; exact fun h ↦ hs hnm h
+  convert h
+  rw [← prod_iUnion, hs_iUnion]
+  simp only [univ_prod_univ, measure_univ]
+
+lemma tendsto_integral_density_of_antitone (hκν : fst κ ≤ ν) [IsFiniteKernel κ]
+    [IsFiniteKernel ν] (a : α) (s : ℕ → Set β) (hs : Antitone s) (hs_iInter : ⋂ i, s i = ∅)
+    (hs_meas : ∀ n, MeasurableSet (s n)) :
+    Tendsto (fun m ↦ ∫ x, density κ ν a x (s m) ∂(ν a)) atTop (𝓝 0) := by
+  simp_rw [integral_density hκν a (hs_meas _)]
+  rw [← ENNReal.zero_toReal]
+  have h_cont := ENNReal.continuousOn_toReal.continuousAt (x := 0) ?_
+  swap
+  · rw [mem_nhds_iff]
+    refine ⟨Iio 1, fun x hx ↦ ne_top_of_lt (?_ : x < 1), isOpen_Iio, ?_⟩
+    · simpa using hx
+    · simp
+  refine h_cont.tendsto.comp ?_
+  have h := tendsto_measure_iInter (s := fun n ↦ univ ×ˢ s n) (μ := κ a)
+    (fun n ↦ MeasurableSet.univ.prod (hs_meas n)) ?_ ?_
+  rotate_left
+  · intro n m hnm x; simp only [mem_prod, mem_univ, true_and]; exact fun h ↦ hs hnm h
+  · refine ⟨0, measure_ne_top _ _⟩
+  convert h
+  rw [← prod_iInter, hs_iInter]
+  simp only [ne_eq, prod_empty, OuterMeasure.empty', forall_exists_index]
+
+lemma tendsto_density_atTop_ae_of_antitone (hκν : fst κ ≤ ν) [IsFiniteKernel κ]
+    [IsFiniteKernel ν] (a : α) (s : ℕ → Set β) (hs : Antitone s) (hs_iInter : ⋂ i, s i = ∅)
+    (hs_meas : ∀ n, MeasurableSet (s n)) :
+    ∀ᵐ x ∂(ν a), Tendsto (fun m ↦ density κ ν a x (s m)) atTop (𝓝 0) := by
+  refine tendsto_of_integral_tendsto_of_antitone ?_ (integrable_const _) ?_ ?_ ?_
+  · exact fun n ↦ integrable_density hκν _ (hs_meas n)
+  · rw [integral_zero]
+    exact tendsto_integral_density_of_antitone hκν a s hs hs_iInter hs_meas
+  · exact ae_of_all _ (fun c n m hnm ↦ density_mono_set hκν a c (hs hnm))
+  · exact ae_of_all _ (fun x m ↦ density_nonneg hκν a x (s m))
+
+section UnivFst
+
+/-! We specialize to `ν = fst κ`, for which `density κ (fst κ) a t univ = 1` almost everywhere. -/
+
+lemma densityProcess_univ [IsFiniteKernel κ] (n : ℕ) (a : α) (x : γ) :
```

Comment:
The naming of this lemma (and of the following ones) is a little bit too generic, as it does not indicate you are considering a specific situation. Add `fst` somewhere, here and below?